Add first draft on working with real numbers (#558)
fpvandoorn/carleson: paper/main.tex

The project differentiated between three different types of (extended) real numbers, $\real$, $\nnreal$ and $\ennreal$,
which gave necessary benefits but also created difficulties in the formalization
and tension in design decisions.
Indeed, in which of the three types to put constants,
subsets, integrands, and perform computations?

A term $x$ of type $\nnreal$ is not also of type $\real$, nor is it of type $\ennreal$, even though in set-based mathematics there are inclusions of the corresponding sets. In type-theory however, one needs coercions, or casts, to go from one type to another. These coercions are often inserted and applied automatically.
The \texttt{norm\_cast} tactic can be really useful to normalize such casts.

However, one also needs conversions in the other directions: sometimes one would need to convert real numbers to nonnegative real numbers. The conversions often play together with other operations, in that one for instance would need to know if, and often under which conditions, the product of the conversions is the conversion of the product. Indeed,
\texttt{
Real.toNNReal (-2) * Real.toNNReal (-3) != Real.toNNReal ((-2) * (-3))

> }.

Reviewer: ```suggestion
\texttt{
Real.toNNReal (-2) * Real.toNNReal (-3) != Real.toNNReal ((-2) * (-3)).
}
```
(to avoid an isolated period).

I would explain here that `Real.toNNReal` is defined to be zero for negative numbers (for those not familiar with Lean/Mathlib).

> Real.toNNReal (-2) * Real.toNNReal (-3) != Real.toNNReal ((-2) * (-3)).

Reviewer: ```suggestion
Real.toNNReal (-2) * Real.toNNReal (-3) ≠ Real.toNNReal ((-2) * (-3)).
```

Even with the necessary side conditions, the \texttt{norm\_cast} tactic does not get very far
in these situations.
This can make conversions between number types quite painful,
and puts some pressure to limit conversions and put as many numbers as possible in the same number type.

Sometimes one is forced to add a conversion, because certain operations are not defined on all types.
Consider for instance the example of performing a computation on exponents in $L^p$-function spaces.
Around the real interpolation theorem, one needs to compute $p$ in terms of other exponents that
naturally live in $\ennreal$.

> Yet there is no definition for $a^p$ if $p$ is in $\ennreal$, and therefore conversions to from

Reviewer: ```suggestion
Yet there is no definition for $a^p$ if $p$ is in $\ennreal$, and therefore conversions from
```

$\ennreal$ to $\real$ are unavoidable.

There are also advantages of differentiating between the spaces.
An advantage of putting for instance a constant in \texttt{NNReal} over \texttt{ENNReal} or
\texttt{Real}, is that it encodes that the constant cannot be infinity and that it needs to be positive.

Another advantage of differentiating comes from the different levels of tactic support.
For real numbers, computations are much easier to perform, using the \texttt{linarith} and \texttt{field\_simp} tactics.
These tactics do not work for terms in $\ennreal$ or $\nnreal$.

> Fortunately, the tactic \texttt{ring} does work for these types.

Reviewer: `ring` works sometimes for these `NNReal` and `ENNReal`, but it is less powerful. For example, goals involving subtraction cannot be handled, as well as some goals involving division or inverses. Here are some examples where `ring` fails (but would succeed over `Real`); it would be good to include some of these in the paper:
```
import Mathlib

open scoped ENNReal NNReal

example (x y : ℝ≥0) : x + y + y - x = 2 * y := by ring
example (x y : ℝ≥0∞) : x + y + y - x = 2 * y := by ring

example (x : ℝ≥0∞) : x / 2 + x / 2 = x := by ring
example (x : ℝ≥0∞) : x⁻¹ ^ 2 = (x ^ 2)⁻¹ := by ring
example (x : ℝ≥0∞) : x⁻¹ ^ 2 = (x ^ 2)⁻¹ := by ring
example (x y : ℝ≥0∞) : (x^2)⁻¹ / y = (y * x^2)⁻¹ := by ring
example (x y : ℝ≥0∞) : 3 / (x - x + y)⁻¹ = 3 * (x + y⁻¹ - x)⁻¹ := by ring
example (x : ℝ≥0∞) : (x : ℝ≥0∞) = x * (5/3)*(3/5) := by ring

example (x y : ℝ≥0) : 3 / (x - x + y)⁻¹ = 3 * (x + y⁻¹ - x)⁻¹ := by ring
example (x : ℝ≥0) : (x : ℝ≥0∞) = x * (5/3)*(3/5) := by ring

```

Reviewer 2: :bulb: TIL, thanks!

> For both $\ennreal$ and $\nnreal$, it will generally fail when a goal needs properties of subtraction: it even fails to prove that $x - x = 0$. If $x \in \nnreal$, the \texttt{ring} tactic can prove that $(x^{-1})^2 = (x ^ 2)^{-1}$, or that $3 / x = 3 x^{-1}$, but it fails if $x \in \ennreal$.

Reviewer: ```suggestion
For both $\ennreal$ and $\nnreal$, it will generally fail when a goal needs properties of subtraction: it even fails to prove that $x - x = 0$. If $x : \nnreal$, the \texttt{ring} tactic can prove that $(x^{-1})^2 = (x ^ 2)^{-1}$, or that $3 / x = 3 x^{-1}$, but it fails if $x : \ennreal$.
```

Reviewer: A related question is whether we should use the `\nnreal` and `\ennreal` notations in this section, or use the Mathlib notations (since we are specifically discussing implementation details). I am not sure which option is better, so I am not asking you to change this (unless someone else suggests this).

> Similarly \texttt{norm\_num} does work for also for numbers in $\ennreal$, but again not as well as for real numbers,

Reviewer 2: ```suggestion
Similarly \texttt{norm\_num} also works for numbers in $\ennreal$, but again not as well as for real numbers,
```
maybe?

Author: perhaps: does also work?

As a final example around considerations on which number types to use,
consider the the supremum of a set.
As one eventually is anyway interested in situations in which this supremum is finite and the set
is bounded from above, one could consider a setup in which the sets considered are subsets of
real numbers and the suprema are real numbers as well.
This was the choice at the outset of the project.
However, in this setup, one needs to constantly carry around a proof or assumption that the
set is indeed bounded.
This is cumbersome, and therefore it turned out to be much better to
let suprema just take values in the \texttt{ENNReal}s.

> During the project, we gradually built up intuition on where to put constants and where

Reviewer: I would add that we reached the conclusion that using `ENNReal`  as often as possible was the best choice (although, as you point out, there are cases where we need to use the other types), and connect this with the `ENorm` section.

(You do write this for suprema, but I think this final sentence slightly undermines this point).
